File: kliktahu/skor.py
"""kliktahu/skor.py - RUMUS SKOR dari prompt pemilik (PROMPT_KLIKTAHU.txt §8) + v7. Fungsi MURNI (tanpa I/O).

Satu implementasi dipakai analisis/v3-v6 dan mesin riset real-time (kliktahu/riset). Port TypeScript identik
ada di skema/ts/src/skor.ts dan diuji PARITAS terhadap fixture yang dihasilkan modul ini
(skema/ts/fixture/skor_paritas.json) - kedua bahasa harus memberi angka yang sama persis.

  v3  skor        = jml + kuat*0.7 + niat*0.9 + sains*3 + yt*0.3
      skor_tumbuh = skor + rel*1.2 + kom*1.5 + vis*0.8 + ever*3
  v4  skor_keluarga = skor_tumbuh + jaring*2.5 + kedalaman*1.5 + peluang*2
  v5  velocity   = (jml - jml_lalu)/max(5, jml_lalu) + 0.5 * porsi_frasa_baru
      momen      = 1 - sisa_hari/(jendela*1.25)   (0 <= sisa_hari <= jendela, jendela 45 hari)
      skor_views = (jml + kuat + yt*2 + vis*0.8 + niat*0.6 + sains*2 + jaring*2) + velocity*4 + momen*6
  v6  papan 0-100 = 34 permintaan + 18 wikipedia + 18 celah + 10 visual + 10 pilar + 10 momen
  v7  PELUANG 0-100 = 100 * sum(bobot_i * x_i) atas 8 komponen real-time (lihat BOBOT_V7), plus
      KEYAKINAN 0..1 = porsi bobot yang benar-benar didukung data (komponen tanpa data = netral 0.5, keyakinan 0).
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Mapping, Sequence
from typing import TypedDict

Nilai = tuple[float, float]  # (x 0..1, keyakinan 0..1)


class Sinyal(TypedDict):
    jml: float
    kuat: float
    niat: float
    yt: float
    sains: float
    rel: float
    kom: float
    vis: float
    ever: float


def _j(x: float) -> float:
    """jepit ke 0..1"""
    return 0.0 if x < 0 else 1.0 if x > 1 else x


# ------------------------------------------------------------------------------------------------ v3
def v3_skor(s: Mapping[str, float]) -> float:
    return s["jml"] + s["kuat"] * 0.7 + s["niat"] * 0.9 + s["sains"] * 3 + s["yt"] * 0.3


def v3_tumbuh(s: Mapping[str, float], skor: float | None = None) -> float:
    skor = v3_skor(s) if skor is None else skor
    return skor + s["rel"] * 1.2 + s["kom"] * 1.5 + s["vis"] * 0.8 + s["ever"] * 3


def kuat(posisi: Iterable[int]) -> float:
    """kekuatan saran: posisi 0 (teratas) bernilai 1.0, posisi 9 bernilai 0.1."""
    return sum((10 - p) / 10 for p in posisi)


# ------------------------------------------------------------------------------------------------ v4
def v4_keluarga(tumbuh: float, jaring: float, kedalaman: float, peluang: float) -> float:
    return tumbuh + jaring * 2.5 + kedalaman * 1.5 + peluang * 2


# ------------------------------------------------------------------------------------------------ v5
def v5_velocity(jml: float, frasa: Sequence[str], jml_lalu: float | None, frasa_lalu: Sequence[str] | None) -> float:
    if jml_lalu is None or frasa_lalu is None:
        return 0.0
    baru = len(set(frasa) - set(frasa_lalu)) / max(1, len(frasa))
    return (jml - jml_lalu) / max(5.0, jml_lalu) + 0.5 * baru


def v5_momen(sisa_hari: float | None, jendela: float = 45) -> float:
    if sisa_hari is None or sisa_hari < 0 or sisa_hari > jendela:
        return 0.0
    return 1.0 - sisa_hari / (jendela * 1.25)


def v5_views(s: Mapping[str, float], jaring: float, velocity: float, momen: float) -> float:
    return (
        (s["jml"] + s["kuat"] + s["yt"] * 2 + s["vis"] * 0.8 + s["niat"] * 0.6 + s["sains"] * 2 + jaring * 2)
        + velocity * 4
        + momen * 6
    )


# ------------------------------------------------------------------------------------------------ v6
def norm01(vals: Sequence[float]) -> list[float]:
    if not vals:
        return []
    lo, hi = min(vals), max(vals)
    return [0.5 if hi - lo < 1e-9 else (v - lo) / (hi - lo) for v in vals]


def skor_wiki(views60: float, tren: float = 1.0) -> float:
    """pageview Wikipedia 60 hari + tren (30 hari terakhir / 30 hari sebelumnya) -> 0..1 (100 rb/60 hari ~ 1)."""
    v = min(1.0, math.log10(max(1.0, views60)) / 5.0)
    tr = _j((tren - 0.7) / 0.9)
    return 0.65 * v + 0.35 * tr


def skor_celah(jumlah: float, umur_median_hari: float = 730, median_views: float = 0) -> float:
    """celah = 1 - kejenuhan pesaing (banyak video, baru, views tinggi = jenuh)."""
    jml = min(1.0, jumlah / 40.0)
    baru = 1.0 - min(1.0, umur_median_hari / 730.0)
    views = min(1.0, median_views / 500000.0)
    return 1.0 - (0.45 * jml + 0.2 * baru + 0.35 * views)


def bobot_pilar(rata: Mapping[str, float], pilar: Sequence[str]) -> dict[str, float]:
    """rata = pilar -> rata-rata (views x retensi). Pilar tanpa data = 0.4; terbaik = 1.0; lainnya 0.25..1."""
    if not rata:
        return {p: 0.5 for p in pilar}
    mx = max(rata.values()) or 1.0
    return {p: (0.25 + 0.75 * rata[p] / mx) if p in rata else 0.4 for p in pilar}


def v6_papan(permintaan: float, wiki: float, celah: float, visual: float, pilar_w: float, momen: float) -> float:
    return 34 * permintaan + 18 * wiki + 18 * celah + 10 * visual + 10 * pilar_w + 10 * momen


# ------------------------------------------------------------------------------------------------ v7
BOBOT_V7: dict[str, float] = {
    "permintaan": 0.26,  # v5 skor_views dinormalisasi (autocomplete Google + YouTube asli)
    "minat": 0.12,  # Wikipedia: tingkat + tren pageview
    "momentum": 0.14,  # NAIK sekarang: velocity snapshot + berita + Google Trends + lonjakan Wikipedia
    "celah": 0.16,  # kejenuhan pesaing YouTube + rasio outlier (views/subscriber)
    "kecocokan": 0.12,  # bisa divisualkan + bobot pilar (performa) + evergreen
    "waktu": 0.10,  # momen (kalender statis + astronomi + umpan live)
    "kesegaran": 0.05,  # belum dibahas & tidak mirip topik lama
    "bukti": 0.05,  # tersedia sumber kredibel (jurnal/lembaga) untuk fakta
}
assert abs(sum(BOBOT_V7.values()) - 1.0) < 1e-12


def momentum(
    velocity: float | None = None,
    berita_rasio: float | None = None,
    tren_traffic: float | None = None,
    wiki_lonjakan: float | None = None,
) -> Nilai | None:
    """gabungan sinyal 'sedang naik'. Tiap bagian 0..1; bobot bagian yang ada dinormalisasi ulang.
    velocity -> 0.5+0.5*tanh(v); berita_rasio (7 hari vs rata-rata mingguan sebelumnya) -> 0.5+0.25*log2(r);
    tren_traffic (Google Trends harian, perkiraan pencarian) -> log10(t)/5; wiki_lonjakan (skor-z 7 hari) -> 0.5+z/6."""
    bagian: list[tuple[float, float]] = []
    if velocity is not None:
        bagian.append((0.35, 0.5 + 0.5 * math.tanh(velocity)))
    if berita_rasio is not None:
        bagian.append((0.25, _j(0.5 + 0.25 * math.log2(max(berita_rasio, 1e-3)))))
    if tren_traffic is not None:
        bagian.append((0.20, _j(math.log10(max(tren_traffic, 1.0)) / 5.0)))
    if wiki_lonjakan is not None:
        bagian.append((0.20, _j(0.5 + wiki_lonjakan / 6.0)))
    if not bagian:
        return None
    w = sum(b[0] for b in bagian)
    return (sum(b[0] * b[1] for b in bagian) / w, w)


def celah_v7(celah: float | None = None, rasio_outlier: float | None = None) -> Nilai | None:
    """celah pesaing + outlier: median(views/subscriber) tinggi = permintaan melebihi pasokan (celah lebih besar)."""
    if celah is None and rasio_outlier is None:
        return None
    s_out = _j(0.5 + 0.25 * math.log10(max(rasio_outlier, 1e-3))) if rasio_outlier is not None else None
    if celah is not None and s_out is not None:
        return (0.7 * celah + 0.3 * s_out, 1.0)
    if celah is not None:
        return (celah, 0.85)
    return (s_out if s_out is not None else 0.5, 0.4)


def kecocokan(visual: float, pilar_w: float, ever: float, ada_performa: bool) -> Nilai:
    return (0.45 * visual + 0.35 * pilar_w + 0.20 * ever, 1.0 if ada_performa else 0.65)


def kesegaran(status: str, mirip_maks: float = 0.0) -> Nilai:
    """'segar' = 1 (dikurangi bila mirip topik lama > 0.6), 'long' = 0.6 (boleh Shorts sudut baru), 'dibahas' = 0."""
    if status == "dibahas":
        return (0.0, 1.0)
    if status == "long":
        return (0.6, 1.0)
    return (max(0.4, 1.0 - max(0.0, mirip_maks - 0.6) * 1.5), 1.0)


def bukti(n_kredibel: int | None) -> Nilai | None:
    return None if n_kredibel is None else (min(1.0, n_kredibel / 4.0), 1.0)


def v7_peluang(
    komponen: Mapping[str, Nilai | None], tanpa: Iterable[str] = ()
) -> tuple[float, float, dict[str, float]]:
    """-> (peluang 0..100, keyakinan 0..1, x per komponen). Komponen tanpa data = netral 0.5, keyakinan 0.
    `tanpa` = komponen yang TIDAK DIPAKAI kanal (mis. 'celah' bila analisis pesaing dimatikan pemilik): bobotnya
    dikeluarkan dan bobot lain dinormalisasi ulang, sehingga keyakinan tidak terkunci oleh sumber yang sengaja
    tidak dipakai. Peringkat tidak berubah (komponen itu netral untuk semua tema); x-nya tetap dilaporkan 0.5."""
    buang = set(tanpa)
    x: dict[str, float] = {}
    total = yakin = wsum = 0.0
    for nama, w in BOBOT_V7.items():
        v = komponen.get(nama)
        xi, ci = (0.5, 0.0) if v is None else (_j(float(v[0])), _j(float(v[1])))
        x[nama] = 0.5 if nama in buang else xi
        if nama in buang:
            continue
        total += w * xi
        yakin += w * ci
        wsum += w
    if not buang:
        return 100.0 * total, yakin, x
    if wsum <= 0:
        return 50.0, 0.0, x
    return 100.0 * total / wsum, yakin / wsum, x


def format_saran(jaring: float, kedalaman: float, ever: float, v7: float) -> str:
    """Long bila pohon pertanyaannya lebar & dalam (cukup untuk ~10 bab) dan evergreen; selain itu Shorts."""
    return "long" if (jaring >= 5 and kedalaman >= 1.6 and ever >= 1 and v7 >= 55) else "shorts"

File: skema/ts/src/skor.ts
/**
 * skema/ts/src/skor.ts - RUMUS SKOR KlikTahu (PROMPT_KLIKTAHU.txt §8 + v7) dalam TypeScript.
 * Port IDENTIK dari kliktahu/skor.py (sumber kebenaran). Diuji paritas terhadap skema/ts/fixture/skor_paritas.json
 * yang dihasilkan Python: kedua bahasa harus memberi angka yang sama (toleransi 1e-9). BUKAN aplikasi - pustaka
 * rumus murni untuk dipakai ulang bila kelak ada integrasi (mis. Bolt Database / supabase-js, lihat tipe.ts).
 *
 *   v3  skor        = jml + kuat*0.7 + niat*0.9 + sains*3 + yt*0.3
 *       skor_tumbuh = skor + rel*1.2 + kom*1.5 + vis*0.8 + ever*3
 *   v4  skor_keluarga = skor_tumbuh + jaring*2.5 + kedalaman*1.5 + peluang*2
 *   v5  velocity   = (jml - jml_lalu)/max(5, jml_lalu) + 0.5 * porsi_frasa_baru
 *       momen      = 1 - sisa_hari/(jendela*1.25)   (0 <= sisa_hari <= jendela)
 *       skor_views = (jml + kuat + yt*2 + vis*0.8 + niat*0.6 + sains*2 + jaring*2) + velocity*4 + momen*6
 *   v6  papan = 34 permintaan + 18 wikipedia + 18 celah + 10 visual + 10 pilar + 10 momen
 *   v7  peluang = 100 * sum(bobot_i * x_i), keyakinan = sum(bobot_i * c_i)
 */

export interface Sinyal {
  jml: number;
  kuat: number;
  niat: number;
  yt: number;
  sains: number;
  rel: number;
  kom: number;
  vis: number;
  ever: number;
}

/** (x 0..1, keyakinan 0..1) */
export type Nilai = readonly [number, number];

export type NamaKomponen =
  | "permintaan" | "minat" | "momentum" | "celah" | "kecocokan" | "waktu" | "kesegaran" | "bukti";

export const BOBOT_V7: Readonly<Record<NamaKomponen, number>> = {
  permintaan: 0.26,
  minat: 0.12,
  momentum: 0.14,
  celah: 0.16,
  kecocokan: 0.12,
  waktu: 0.10,
  kesegaran: 0.05,
  bukti: 0.05,
};
const URUT_V7: readonly NamaKomponen[] = [
  "permintaan", "minat", "momentum", "celah", "kecocokan", "waktu", "kesegaran", "bukti",
];

const jepit = (x: number): number => (x < 0 ? 0 : x > 1 ? 1 : x);

// ------------------------------------------------------------------------------------------------ v3
export function v3Skor(s: Sinyal): number {
  return s.jml + s.kuat * 0.7 + s.niat * 0.9 + s.sains * 3 + s.yt * 0.3;
}

export function v3Tumbuh(s: Sinyal, skor: number | null = null): number {
  const sk = skor === null ? v3Skor(s) : skor;
  return sk + s.rel * 1.2 + s.kom * 1.5 + s.vis * 0.8 + s.ever * 3;
}

export function kuat(posisi: readonly number[]): number {
  let t = 0;
  for (const p of posisi) t += (10 - p) / 10;
  return t;
}

// ------------------------------------------------------------------------------------------------ v4
export function v4Keluarga(tumbuh: number, jaring: number, kedalaman: number, peluang: number): number {
  return tumbuh + jaring * 2.5 + kedalaman * 1.5 + peluang * 2;
}

// ------------------------------------------------------------------------------------------------ v5
export function v5Velocity(
  jml: number, frasa: readonly string[], jmlLalu: number | null, frasaLalu: readonly string[] | null,
): number {
  if (jmlLalu === null || frasaLalu === null) return 0;
  const lalu = new Set(frasaLalu);
  const baruSet = new Set(frasa.filter((f) => !lalu.has(f)));
  const baru = baruSet.size / Math.max(1, frasa.length);
  return (jml - jmlLalu) / Math.max(5, jmlLalu) + 0.5 * baru;
}

export function v5Momen(sisaHari: number | null, jendela = 45): number {
  if (sisaHari === null || sisaHari < 0 || sisaHari > jendela) return 0;
  return 1 - sisaHari / (jendela * 1.25);
}

export function v5Views(s: Sinyal, jaring: number, velocity: number, momen: number): number {
  return (s.jml + s.kuat + s.yt * 2 + s.vis * 0.8 + s.niat * 0.6 + s.sains * 2 + jaring * 2) + velocity * 4 + momen * 6;
}

// ------------------------------------------------------------------------------------------------ v6
export function norm01(vals: readonly number[]): number[] {
  if (vals.length === 0) return [];
  const lo = Math.min(...vals);
  const hi = Math.max(...vals);
  return vals.map((v) => (hi - lo < 1e-9 ? 0.5 : (v - lo) / (hi - lo)));
}

export function skorWiki(views60: number, tren = 1.0): number {
  const v = Math.min(1, Math.log10(Math.max(1, views60)) / 5);
  const tr = jepit((tren - 0.7) / 0.9);
  return 0.65 * v + 0.35 * tr;
}

export function skorCelah(jumlah: number, umurMedianHari = 730, medianViews = 0): number {
  const jml = Math.min(1, jumlah / 40);
  const baru = 1 - Math.min(1, umurMedianHari / 730);
  const views = Math.min(1, medianViews / 500000);
  return 1 - (0.45 * jml + 0.2 * baru + 0.35 * views);
}

export function bobotPilar(rata: Readonly<Record<string, number>>, pilar: readonly string[]): Record<string, number> {
  const kunci = Object.keys(rata);
  const out: Record<string, number> = {};
  if (kunci.length === 0) {
    for (const p of pilar) out[p] = 0.5;
    return out;
  }
  const mx = Math.max(...kunci.map((k) => rata[k] ?? 0)) || 1;
  for (const p of pilar) {
    const r = rata[p];
    out[p] = r !== undefined ? 0.25 + (0.75 * r) / mx : 0.4;
  }
  return out;
}

export function v6Papan(
  permintaan: number, wiki: number, celah: number, visual: number, pilarW: number, momen: number,
): number {
  return 34 * permintaan + 18 * wiki + 18 * celah + 10 * visual + 10 * pilarW + 10 * momen;
}

// ------------------------------------------------------------------------------------------------ v7
export function momentum(
  velocity: number | null = null, beritaRasio: number | null = null, trenTraffic: number | null = null,
  wikiLonjakan: number | null = null,
): Nilai | null {
  const bagian: Array<[number, number]> = [];
  if (velocity !== null) bagian.push([0.35, 0.5 + 0.5 * Math.tanh(velocity)]);
  if (beritaRasio !== null) bagian.push([0.25, jepit(0.5 + 0.25 * Math.log2(Math.max(beritaRasio, 1e-3)))]);
  if (trenTraffic !== null) bagian.push([0.20, jepit(Math.log10(Math.max(trenTraffic, 1)) / 5)]);
  if (wikiLonjakan !== null) bagian.push([0.20, jepit(0.5 + wikiLonjakan / 6)]);
  if (bagian.length === 0) return null;
  let w = 0;
  for (const b of bagian) w += b[0];
  let t = 0;
  for (const b of bagian) t += b[0] * b[1];
  return [t / w, w];
}

export function celahV7(celah: number | null = null, rasioOutlier: number | null = null): Nilai | null {
  if (celah === null && rasioOutlier === null) return null;
  const sOut = rasioOutlier !== null ? jepit(0.5 + 0.25 * Math.log10(Math.max(rasioOutlier, 1e-3))) : null;
  if (celah !== null && sOut !== null) return [0.7 * celah + 0.3 * sOut, 1];
  if (celah !== null) return [celah, 0.85];
  return [sOut ?? 0.5, 0.4];
}

export function kecocokan(visual: number, pilarW: number, ever: number, adaPerforma: boolean): Nilai {
  return [0.45 * visual + 0.35 * pilarW + 0.20 * ever, adaPerforma ? 1 : 0.65];
}

export function kesegaran(status: string, miripMaks = 0): Nilai {
  if (status === "dibahas") return [0, 1];
  if (status === "long") return [0.6, 1];
  return [Math.max(0.4, 1 - Math.max(0, miripMaks - 0.6) * 1.5), 1];
}

export function bukti(nKredibel: number | null): Nilai | null {
  return nKredibel === null ? null : [Math.min(1, nKredibel / 4), 1];
}

/** tanpa = komponen yang TIDAK DIPAKAI kanal (mis. "celah" bila analisis pesaing dimatikan): bobotnya dikeluarkan
 *  dan bobot lain dinormalisasi ulang (paritas persis dengan kliktahu/skor.py v7_peluang). */
export function v7Peluang(
  komponen: Readonly<Partial<Record<NamaKomponen, Nilai | null>>>,
  tanpa: readonly string[] = [],
): [number, number, Record<NamaKomponen, number>] {
  const buang = new Set(tanpa);
  const x = {} as Record<NamaKomponen, number>;
  let total = 0;
  let yakin = 0;
  let wsum = 0;
  for (const nama of URUT_V7) {
    const w = BOBOT_V7[nama];
    const v = komponen[nama] ?? null;
    const xi = v === null ? 0.5 : jepit(v[0]);
    const ci = v === null ? 0 : jepit(v[1]);
    x[nama] = buang.has(nama) ? 0.5 : xi;
    if (buang.has(nama)) continue;
    total += w * xi;
    yakin += w * ci;
    wsum += w;
  }
  if (buang.size === 0) return [100 * total, yakin, x];
  if (wsum <= 0) return [50, 0, x];
  return [(100 * total) / wsum, yakin / wsum, x];
}

export function formatSaran(jaring: number, kedalaman: number, ever: number, v7: number): "long" | "shorts" {
  return jaring >= 5 && kedalaman >= 1.6 && ever >= 1 && v7 >= 55 ? "long" : "shorts";
}
